radix-engine/src/utils/native_blueprint_call_validator.rs

```diff
@@ -0,0 +1,460 @@
+            EntityType::InternalFungibleVault => package_definitions
+                .resource_package()
+                .schema
+                .blueprints
+                .get(FUNGIBLE_VAULT_BLUEPRINT),
+            EntityType::InternalNonFungibleVault => package_definitions
+                .resource_package()
+                .schema
+                .blueprints
+                .get(NON_FUNGIBLE_VAULT_BLUEPRINT),
+
+            EntityType::GlobalGenericComponent
+            | EntityType::InternalGenericComponent
+            | EntityType::InternalKeyValueStore => None,
+        },
+        Invocation::Method(_, ObjectModuleId::Metadata, _) => package_definitions
+            .metadata_package()
+            .schema
+            .blueprints
+            .get(METADATA_BLUEPRINT),
+        Invocation::Method(_, ObjectModuleId::AccessRules, _) => package_definitions
+            .access_rules_package()
+            .schema
+            .blueprints
+            .get(ACCESS_RULES_BLUEPRINT),
+        Invocation::Method(_, ObjectModuleId::Royalty, _) => package_definitions
+            .royalties_package()
+            .schema
+            .blueprints
+            .get(COMPONENT_ROYALTY_BLUEPRINT),
+    };
+
+    if let Some(blueprint_schema) = blueprint_schema {
+        if let Some(function_schema) = blueprint_schema.functions.get(invocation.method()) {
+            if function_schema.receiver.is_none() && invocation.is_function()
+                || function_schema.receiver.is_some() && invocation.is_method()
+            {
+                Ok(Some((function_schema.input, &blueprint_schema.schema)))
+            } else {
+                Err(ValidationError::InvalidReceiver(instruction_index))
+            }
+        } else {
+            Err(ValidationError::MethodNotFound(
+                instruction_index,
+                invocation.method().to_owned(),
+            ))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+enum Invocation {
+    Method(GlobalAddress, ObjectModuleId, String),
+    Function(PackageAddress, String, String),
+}
+
+impl Invocation {
+    fn method(&self) -> &str {
+        match self {
+            Self::Method(_, _, method) => method,
+            Self::Function(_, _, method) => method,
+        }
+    }
+
+    fn global_address(&self) -> GlobalAddress {
+        match self {
+            Self::Method(global_address, ..) => *global_address,
+            Self::Function(package_address, ..) => (*package_address).into(),
+        }
+    }
+
+    fn is_function(&self) -> bool {
+        match self {
+            Self::Function(..) => true,
+            Self::Method(..) => false,
+        }
+    }
+
+    fn is_method(&self) -> bool {
+        match self {
+            Self::Method(..) => true,
+            Self::Function(..) => false,
+        }
+    }
+}
+
+// `usize` here is the index of the instruction has has an error.
+#[derive(Clone, Debug)]
+pub enum ValidationError {
```

Comment:
This is exposed as `radix_engine::utils::ValidationError` which doesn't feel like quite the right description.

Perhaps `ManifestSchemaValidationError`? Or else put this in `radix_engine::utils::schema::ValidationError` instead?